### ISSUE TITLE
CoinJoin Implementation

### ISSUE THREAD

**Issue by christroutner** [Bitcoin-com/bitbox-javascript-sdk#26]
**As a user of Bitcoin Cash, I would like to engage in private transactions using CoinJoin or some other privacy enabling protocol.**

The scope of this issue is to implement CoinJoin or an algorithm like it into BITBOX. This is also an opportunity to open up a conversation about transaction privacy and ways to implement it in Bitcoin Cash and BITBOX.

## Reference
The original CoinJoin post:
https://bitcointalk.org/index.php?topic=279249

Two BTC wallets that implement CoinJoin:
* [JoinMarket](https://github.com/JoinMarket-Org/joinmarket/wiki) ([Tails installer](https://github.com/JoinMarket-Org/tailsjoin))
* [Wasabi Wallet](https://github.com/zkSNACKs/WalletWasabi)

**Comment by christroutner** [comment]
Please discuss this idea in the #bitbox-sdk channel on Discord:
https://discord.gg/m5h28ND

But post to this Issue for any links or information that needs to be retained over the long-term.

**Comment by christroutner** [comment]
I found this JavaScript CoinJoin library created by the Dark Wallet project. It's a little dated. Does anyone know any background on this? Is it secure? Accurate? Anyone used it?
https://github.com/darkwallet/coinjoin.js

**Comment by HostFat** [comment]
There is also this https://github.com/cashshuffle/cashshuffle-electron-cash-plugin

**Comment by imaginaryusername** [comment]
Most coinjoin protocols, including Joinmarket, suffer from the fact that users have to shuffle before using in a separate process, whether actively or automated. Joinmarket itself can facilitate arbitrary amounts, but still has the problem where it leaves behind change that can be linked to "maker" status by fees, making it easier to track payments if you compromise the merchant. In Bitcoin Cash since small amounts of fees can be separated into their own UTXO and remain usable, I propose to separate them as an improvement over Joinmarket - and the implementation needs to be in light wallets as opposed to the strict full-nodes-only-for-makers policy in Joinmarkets. Hope this is of some use. https://github.com/imaginaryusername/payshuffle/blob/master/Payshuffle.pdf

**Comment by christroutner** [comment]
I skimmed through the PDF. I'm really looking forward to diving into this. Thank you for the quality information. 

Thanks to [BITBOX HD Wallets library](https://github.com/Bitcoin-com/bitbox-javascript-sdk/tree/master/examples), I think we could eliminate any requirements for a full node.

**Comment by christroutner** [comment]
Just came across this CashShuffle JavaScript library:
https://github.com/bookmoons/cashshufflejs

**Comment by christroutner** [comment]
I spent some time yesterday talking with @zquestz and learning about CashShuffle. Very interesting!

Here is the breakdown: It's a server-client model. The server is really dumb, and a bulk of the work is done by the clients. Messages are passed using proto-bufs, which have the effect of not tracking peoples IP addresses.

The client is written in Python and is an Electron Cash app:
https://github.com/cashshuffle/cashshuffle-electron-cash-plugin

This client could be ported over to JavaScript and BITBOX SDK would be a perfect application for it. But this would be a major undertaking.

There is already a $5K USD fund to do this work, sponsored by the Bitcoin.com Wallet team:
https://github.com/Bitcoin-com/Wallet/issues/34

**Comment by BobWallets** [comment]
I have a working implementation

https://github.com/BobWallet/BobWallet

**Comment by christroutner** [comment]
Wow! That BobWallet is. so. awesome. Thank you for sharing. I will be playing around with it and looking through the code!

I had this idea for an anonymizing protocol the last few days, and I refined it to a point where I couldn't see any glaring holes. I'd love for people to review it and leave comments to solicit feedback.
https://gist.github.com/christroutner/5ba3b22cc9ae9f64b70f1004bd551cbf

**Comment by christroutner** [comment]
Closing this issue as it's been open long enough to get significant discussion.

CashShuffle seems to be taking the lead in interest in community development. Here is an article I wrote on how to setup the Electron Cash client:
https://honest.cash/christroutner/cashshuffle-how-to-install-and-test-2083

They will be releasing a JavaScript library soon: 
https://github.com/clifordsymack/cashshufflejs

which will make it much easier to integrate in Badger Wallet and my [bch-cli-wallet](https://github.com/christroutner/bch-cli-wallet)

Another protocol that I will continue to work on is Consolidating CoinJoin. I have a series of articles on Honest.cash discussing the technologies I plan to put together for it, including a video demo of the software for it:
https://honest.cash/christroutner/illegal-code-102
